### PR TITLE
Support PIL/ndarray as input

### PR DESCRIPTION
**Pull request by Laughing-q**
@AyushExel @glenn-jocher 
usage:
```python
# from PIL/ndarray, we just purely return outputs, no image/video saving and showing.
from ultralytics import YOLO
from PIL import Image
import cv2

model = YOLO("model.pt")
img = cv2.imread("path/img")
outputs = model.predict(img=img)  # accepts opencv/ndarray
img = Image.open("path/img")
outputs = model.predict(img=img)  # accepts PIL

# also accepts multiple images(batch inference)
# support different resolution images in one list(pt model only).
outputs = model.predict(img=[img, img])  # accepts List, 
```

also I updated `test_python.py` and Docs about this part. And passed all the tests.
image
image

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced Python interface for more flexible image prediction with Ultralytics YOLO models.

### 📊 Key Changes
- Simplified Python use-case examples in the documentation.
- Added functionality to perform predictions on images supplied either as a PIL Image, NumPy ndarray, or a list of images for batch inference.
- Overhauled the internal logic to support image input directly, rather than just file paths.
- Introduced clear distinctions between different source inputs (`source` parameter vs. `img` parameter).
- Implemented tests to confirm functionality with PIL Image and NumPy ndarray inputs, as well as ensuring batch prediction works as expected.
- Refined the AI model's call behavior to be more intuitive and cover more use cases.

### 🎯 Purpose & Impact
- Enables developers and users to easily apply YOLO model predictions to images stored in memory, rather than just those saved to disk. 🛠️
- Provides a straightforward way of performing batch inferences, which can significantly speed up the prediction process when dealing with multiple images. 🚀
- Improves user experience by offering an intuitive and flexible Python API for various prediction tasks, potentially broadening the user base of the Ultralytics YOLO models. 🌍
- Establishes a robust test suite for new functionalities, increasing the reliability and stability of the software. ✅

### CODE REVIEW

File: ultralytics/yolo/engine/predictor.py [hunk 1]
```diff
@@ -220,6 +217,51 @@ def predict_cli(self, source=None, model=None, return_outputs=False):
...
+        auto = all(x.shape == img[0].shape for x in img) and self.model.pt
+        im = [self._single_preprocess(im, self.model.stride, auto) for im in img]
+        im = np.stack(im, 0) if len(im) > 1 else im[0]
+
+        im = self.preprocess(im)
+        if len(im.shape) == 3:
+            im = im[None]  # expand for batch dim
+        preds = self.model(im)
+        preds = self.postprocess(preds, im, img)
+        return preds
```
Comment (Laughing-q) [inline ultralytics/yolo/engine/predictor.py:245]: @AyushExel ideally once your `Results` class is done, you can just update this line to:
```python
return Result(preds)
```

Comment (AyushExel) [reply ultralytics/yolo/engine/predictor.py:245]: yup that sounds good!

**Comment by kadirnar** [comment]
When will this pr be accepted?

**Comment by AyushExel** [comment]
@Laughing-q you've merged this PR into the results PR right? So this one can be closed?

**Comment by Laughing-q** [comment]
@AyushExel yes

**Pull request closed without merging**